/**
 * \file mergerControlWidget.cpp
 * \attention This file is automatically generated and should not be in general modified manually
 *
 * \date MMM DD, 20YY
 * \author autoGenerator
 */

#include "mergerControlWidget.h"
#include "ui_mergerControlWidget.h"
#include <memory>
#include "qSettingsGetter.h"
#include "qSettingsSetter.h"


MergerControlWidget::MergerControlWidget(QWidget *parent, bool _autoInit, QString _rootPath)
    : ParametersControlWidgetBase(parent)
    , mUi(new Ui::MergerControlWidget)
    , autoInit(_autoInit)
    , rootPath(_rootPath)
{
    mUi->setupUi(this);

    QObject::connect(mUi->undistCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->fOVSpinBox, SIGNAL(valueChanged(double)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->outSizeSpinBox, SIGNAL(valueChanged(double)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->outPhySizeSpinBox, SIGNAL(valueChanged(double)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->groundZSpinBox, SIGNAL(valueChanged(double)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->switch1CheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->switch2CheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->switch3CheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->switch4CheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
}

MergerControlWidget::~MergerControlWidget()
{

    delete mUi;
}

void MergerControlWidget::loadParamWidget(WidgetLoader &loader)
{
    std::unique_ptr<Merger> params(createParameters());
    loader.loadParameters(*params, rootPath);
    setParameters(*params);
}

void MergerControlWidget::saveParamWidget(WidgetSaver  &saver)
{
    saver.saveParameters(*std::unique_ptr<Merger>(createParameters()), rootPath);
}

void MergerControlWidget::getParameters(Merger& params) const
{
    params = *std::unique_ptr<Merger>(createParameters());
}


Merger *MergerControlWidget::createParameters() const
{

    /**
     * We should think of returning parameters by value or saving them in a preallocated place
     **/


    return new Merger(
          mUi->undistCheckBox->isChecked()
        , mUi->fOVSpinBox->value()
        , mUi->outSizeSpinBox->value()
        , mUi->outPhySizeSpinBox->value()
        , mUi->groundZSpinBox->value()
        , mUi->switch1CheckBox->isChecked()
        , mUi->switch2CheckBox->isChecked()
        , mUi->switch3CheckBox->isChecked()
        , mUi->switch4CheckBox->isChecked()
    );
}

void MergerControlWidget::setParameters(const Merger &input)
{
    // Block signals to send them all at once
    bool wasBlocked = blockSignals(true);
    mUi->undistCheckBox->setChecked(input.undist());
    mUi->fOVSpinBox->setValue(input.fOV());
    mUi->outSizeSpinBox->setValue(input.outSize());
    mUi->outPhySizeSpinBox->setValue(input.outPhySize());
    mUi->groundZSpinBox->setValue(input.groundZ());
    mUi->switch1CheckBox->setChecked(input.switch1());
    mUi->switch2CheckBox->setChecked(input.switch2());
    mUi->switch3CheckBox->setChecked(input.switch3());
    mUi->switch4CheckBox->setChecked(input.switch4());
    blockSignals(wasBlocked);
    emit paramsChanged();
}

void MergerControlWidget::setParametersVirtual(void *input)
{
    // Modify widget parameters from outside
    Merger *inputCasted = static_cast<Merger *>(input);
    setParameters(*inputCasted);
}
